crates/pallets/starknet/src/transaction_validation.rs

```diff
@@ -50,3 +53,67 @@ impl<OuterOrigin: Into<Result<RawOrigin, OuterOrigin>> + From<RawOrigin>> Ensure
         Ok(OuterOrigin::from(RawOrigin::StarknetTransaction))
     }
 }
+
+impl<T: Config> Pallet<T> {
+    pub fn validate_usigned_tx_nonce(
+        transaction: &UserAndL1HandlerTransaction,
+    ) -> Result<(Felt252Wrapper, Felt252Wrapper, Felt252Wrapper), InvalidTransaction> {
+        let (sender_address, transaction_nonce) = match transaction {
+            UserAndL1HandlerTransaction::User(tx) => (tx.sender_address().into(), *tx.nonce()),
+            UserAndL1HandlerTransaction::L1Handler(tx, _fee) => (ContractAddress::default(), tx.nonce.into()),
```

Comment:
@tdelabro I need to confirm if `sender_address` for `L1Handler` should be `ContractAddress::default()` or sth else

Reply:
Like here -> https://github.com/keep-starknet-strange/madara/blob/main/crates/pallets/starknet/src/tests/l1_message.rs#L20-L22

Reply:
Good catch. The starknet specs don't enforce ordering for messaging. The nonce only exists for deduplication, not for ordering.

So there is no need for an account to be used. But it also means we have to keep multiple nonces in memory.

I think we should use a persistent hashset in the storage for now.

Check the nonce is not in during validation, put the nonce in during execution.

Reply:
UserTx and messages have to be treated differently there

Reply:
@tdelabro do I understand correctly that:
- I should add HashSet to `Starknet Pallet` for tracking nonces
- I should move nonce check to the execution part
- I should allow execution of messages only with unused nonce (no need for ordering)

Besides that, could you elaborate more on 
> UserTx and messages have to be treated differently there